lib/ros_value.h

```diff
@@ -349,8 +375,9 @@ class RosValue {
     object_info_t object_info_;
   };
 
-  const char* const getPrimitivePointer() const {
-    return &primitive_info_.message_buffer->at(primitive_info_.offset);
+  template<typename T>
+  const T* const getPrimitivePointer() const {
+    return reinterpret_cast<const T* const>(&primitive_info_.message_buffer->at(primitive_info_.offset));
```

Comment:
This looks like it can never be null.
Could you consider returning a reference here instead of a pointer in order to signal that to the user?
Maybe rename it `getPrimitive()`

Reply:
Alternatively, you need to be careful to `assert (ptr != nullptr)` everywhere this is called and not dereference it before it's been checked for null, which is not currently happening

Reply:
Ya you are correct, it can never be null, I can change this to just be getPrimitive (or maybe getPrimitiveReference), as its not too inconvenient to just add the & symbol when you want the pointer.

Reply:
I don't think I understand where this could possibly return nullptr, but maybe I'm missing something, here's my thinking, let me know where I'm tripping up so I can understand - 
`primitive_info_.message_buffer->at(primitive_info_.offset)` will never return a reference to a null object, it will throw an exception if at() points to a bad index. Which means `&primitive_info_.message_buffer->at(primitive_info_.offset)` will never be nullptr, which means `reinterpret_cast<const T* const>(&primitive_info_.message_buffer->at(primitive_info_.offset))`. I don't think I understand how arrays or objects would change this though, so maybe I'm missing something, could you expand there?


In general, I think if you want to keep this function returning a pointer, then you do need to assert that the pointer is not null everywhere that you call this function. Guarantees are held by assertion, not by assumption - you should never assume that a pointer returned from a function will not be null, you should only assert it. Generally in C++, a function that will never return a null pointer would signal that to the caller by returning a reference, not a pointer. If the function returns a pointer, as opposed to a reference, that means the interface is signaling to the reader that nullptr might be returned, which means if you think you're using it in a way where nullptr won't be returned, you should assert that it didn't return nullptr, even if it's a private function - if the private function was guaranteed to not be nullptr, then it should return a reference, whether or not it's private

Reply:
You're totally right - some form of an error will likely get thrown, or the data will just be garbage.
Thanks for the explanation about the pointer vs reference - will update!